src/docbuild/utils/pidlock.py

```diff
@@ -5,165 +5,169 @@
+    def __exit__(self, exc_type, exc_value, traceback) -> None:
         self.release()
 
     def _generate_lock_name(self) -> Path:
-        """Generate a unique lock file name based on the resource path."""
-        # SHA256 hash of the absolute path is used to ensure a unique, safe filename.
-        path_hash = hashlib.sha256(str(self.resource_path).encode('utf-8')).hexdigest()
-        return self._lock_dir / f'docbuild-{path_hash}.pid'
+        path_hash = hashlib.sha256(str(self.resource_path).encode("utf-8")).hexdigest()
+        return self._lock_dir / f"docbuild-{path_hash}.pid"
 
     def acquire(self) -> None:
         """Acquire the lock atomically, or diagnose an existing lock."""
         self.lock_dir.mkdir(parents=True, exist_ok=True)
         current_pid = os.getpid()
 
-        # 1. Attempt to acquire the lock file atomically (O_CREAT | os.O_EXCL)
-        try:
-            # os.O_WRONLY | os.O_CREAT | os.O_EXCL ensures file is created ONLY if it doesn't exist.
-            # 0o644 sets the file permissions (read/write for owner, read for others).
-            lock_fd = os.open(
-                self.lock_path, os.O_WRONLY | os.O_CREAT | os.O_EXCL, 0o644
-            )
-            # Write PID to the atomically created file
-            with os.fdopen(lock_fd, 'w') as f:
-                f.write(str(current_pid) + '\n')
-
-            self._lock_acquired = True
-            log.debug(f"Acquired lock for {self.resource_path} at PID {current_pid}.")
-            # Registering release ensures the lock is cleaned up when the program exits.
-            atexit.register(self.release)
-            return
-
-        except FileExistsError:
-            # 2. Lock file exists (atomic open failed). Must now check if it's stale or running.
-            pass
-        except OSError as e:
-            # Catch file system errors during the atomic open operation
-            raise RuntimeError(f"Failed to create lock file at {self.lock_path}: {e}")
-
-        # 3. Lock exists. Check if the process is running.
-        if self.lock_path.exists():
+        while True:
+            # Step 1: Try atomic creation
             try:
-                # Read PID from the file content (slower operation, only done on contention)
-                with self.lock_path.open('r') as f:
+                lock_fd = os.open(
+                    self.lock_path, os.O_WRONLY | os.O_CREAT | os.O_EXCL, 0o644
+                )
+                with os.fdopen(lock_fd, "w") as f:
+                    f.write(f"{current_pid}\n")
+                self._lock_acquired = True
+                self.lock_file = self.lock_path
+                log.debug(f"Acquired lock for {self.resource_path} at PID {current_pid}.")
+                atexit.register(self.release)
+                return
+            except FileExistsError:
+                pass
+            except OSError as e:
+                raise RuntimeError(f"Failed to create lock file at {self.lock_path}: {e}")
```

Comment:
I think the error message misses the point. The lock is just a technical detail. For _users_ it's important to let them know that the same instance is already running and what to do.

The following is just a suggestion, feel free to adapt it:


```suggestion
                raise RuntimeError(
                    "Same instance is already running. "
                    "Either wait until the first instance is complete or "
                    "cancel it. "
                    f"PID={current_pid}, lock file at {self.lock_path}: {e}"
                    )
```